includes/installer/class-installer.php

```diff
@@ -0,0 +1,150 @@
+<?php
+/**
+ * File containing the class \Sensei\Installer\Installer.
+ *
+ * @package sensei
+ * @since $$next-version$$
+ */
+
+namespace Sensei\Installer;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+ * Installer class.
+ *
+ * Responsible for running DB updates that need to be run per version.
+ *
+ * @since $$next-version$$
+ */
+class Installer {
+	/**
+	 * Instance of the class.
+	 *
+	 * @since $$next-version$$
+	 * @var self
+	 */
+	private static $instance;
+
+	/**
+	 * The database schema class.
+	 *
+	 * @since $$next-version$$
+	 * @var Schema
+	 */
+	private $schema;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since $$next-version$$
+	 *
+	 * @param Schema $schema Schema migration object.
+	 */
+	public function __construct( Schema $schema ) {
+		$this->schema = $schema;
+	}
+
+	/**
+	 * Fetches an instance of the class.
+	 *
+	 * @since $$next-version$$
+	 *
+	 * @return self
+	 */
+	public static function instance(): self {
+		if ( ! self::$instance ) {
+			self::$instance = new self( new Schema() );
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Initialize necessary hooks.
+	 *
+	 * @since $$next-version$$
+	 */
+	public function init() {
+		register_activation_hook( SENSEI_LMS_PLUGIN_FILE, [ $this, 'install' ] );
+		add_action( 'plugins_loaded', [ $this, 'install' ] );
+	}
+
+	/**
+	 * Run the installer.
+	 *
+	 * This method is executed when the plugin is installed or updated.
+	 *
+	 * @since $$next-version$$
+	 */
+	public function install() {
+		if (
+			! is_blog_installed()
+			|| $this->is_installing()
+			|| ! $this->requires_install()
+		) {
+			return;
+		}
+
+		set_transient( 'sensei_lms_installing', 'yes', MINUTE_IN_SECONDS * 10 );
+
+		$this->schema->create_tables();
+		$this->update_version();
+
+		delete_transient( 'sensei_lms_installing' );
+
+		/**
+		 * Fires after the installation completes.
+		 *
+		 * @since $$next-version$$
+		 */
+		do_action( 'sensei_lms_installed' );
+	}
+
+	/**
+	 * Get the Schema instance.
+	 *
+	 * @since $$next-version$$
+	 *
+	 * @return Schema
+	 */
+	public function get_schema(): Schema {
+		return $this->schema;
+	}
+
+	/**
+	 * Check if the installer is running.
+	 *
+	 * @since $$next-version$$
+	 *
+	 * @return bool
+	 */
+	private function is_installing(): bool {
+		return 'yes' === get_transient( 'sensei_lms_installing' );
+	}
+
+	/**
+	 * Determine if the installer needs to be run by checking the plugin's version.
+	 *
+	 * @since $$next-version$$
+	 *
+	 * @return bool
+	 */
+	private function requires_install(): bool {
+		$version = get_option( 'sensei_lms_version' );
+
+		return version_compare( $version, SENSEI_LMS_VERSION, '<' );
+	}
+
+	/**
+	 * Update the plugin's version to the current one.
+	 *
+	 * @since $$next-version$$
+	 */
+	private function update_version() {
+		update_option( 'sensei_lms_version', SENSEI_LMS_VERSION );
+	}
```

Comment:
Generally, I think we must use the same `sensei-verson`.
However, there is a question: who should update it?

Reply:
Looking at `Sensei_Main`, it takes so many responsibilities.
Maybe we can transfer some of them (related to install/update) to our new installer? (Actually, that's what you suggest in the first option.)
I'm reading docs and looking for an implementation.

Reply:
Sounds good! Let me know if I can help with something!

Reply:
Hey @m1r0! I updated the PR with one solution.
Still not sure if it is good.
I tried to avoid adding "unnecessary dependencies" to the Installer, so I added a factory for Sensei_Updates. It also removed some code duplication between Installer and Sensei_Main.
What I don't like now is the "Installer" name as it also runs updates :) Or is it okay? (I mean we run our updates too :)) 
And a question: what do you think, is it possible to rewrite existing Sensei_Updates using our upcoming data migrator? Looks like those updates exist exactly for that.

Reply:
> What I don't like now is the "Installer" name as it also runs updates :) Or is it okay?

I think it's fine. You should check what Woo is [doing](https://github.com/woocommerce/woocommerce/blob/35a4d1ca320156c13db3913806a91605c6fa2da3/plugins/woocommerce/includes/class-wc-install.php#L221) in its [install](https://github.com/woocommerce/woocommerce/blob/35a4d1ca320156c13db3913806a91605c6fa2da3/plugins/woocommerce/includes/class-wc-install.php#L361-L360) class. 🙂 

> And a question: what do you think, is it possible to rewrite existing Sensei_Updates using our upcoming data migrator? Looks like those updates exist exactly for that.

It looks doable. We might have to tweak how the background jobs are called. I've noticed that some updates are calling a background job, but the migrations are also running in the background, so we will have to check how that will work.